pkg/config/dynamicloader.go

```diff
@@ -0,0 +1,495 @@
+package config
+
+import (
+	"context"
+	"errors"
+	"flag"
+	"fmt"
+	"io"
+	"io/fs"
+	"io/ioutil"
+	"net/url"
+	"path/filepath"
+	"strings"
+
+	v2 "github.com/grafana/agent/pkg/integrations/v2"
+	"github.com/grafana/agent/pkg/logs"
+	"github.com/grafana/agent/pkg/metrics"
+	"github.com/grafana/agent/pkg/metrics/instance"
+	"github.com/grafana/agent/pkg/traces"
+	"github.com/hairyhenderson/go-fsimpl"
+	"github.com/hairyhenderson/go-fsimpl/blobfs"
+	"github.com/hairyhenderson/go-fsimpl/filefs"
+	"github.com/hairyhenderson/gomplate/v3/data"
+	"github.com/hairyhenderson/gomplate/v3/loader"
+	"github.com/hashicorp/go-multierror"
+	"github.com/weaveworks/common/server"
+	"gopkg.in/yaml.v2"
+)
+
+// DynamicLoader is used to load configs from a variety of sources and squash them together.
+// This is used by the dynamic configuration feature to load configurations from a set of templates and then run them through
+// gomplate producing an end result.
+type DynamicLoader struct {
+	loader *loader.ConfigLoader
+	mux    fsimpl.FSMux
+	cfg    *LoaderConfig
+}
+
+// NewDynamicLoader instantiates a new DynamicLoader
+func NewDynamicLoader() (*DynamicLoader, error) {
+
+	return &DynamicLoader{
+		loader: nil,
+		mux:    newFSProvider(),
+		cfg:    nil,
```

Comment:
nit: avoid explicitly setting fields to their zero value to keep things a little tidier